packages/@lwc/engine/src/framework/vm.ts

```
@@ -442,70 +416,32 @@ function runDisconnectedCallback(vm: VM) {
-        if (!isUndefined(elm)) {
-            const childVM = getCustomElementVM(elm as HTMLElement);
-            resetComponentStateWhenRemoved(childVM);
-        }
-    }
-}
-
-function runLightChildNodesDisconnectedCallback(vm: VM) {
-    if (process.env.NODE_ENV !== 'production') {
-        assert.isTrue(vm && 'cmpRoot' in vm, `${vm} is not a vm.`);
-    }
-    const { aChildren: adoptedChildren } = vm;
-    recursivelyDisconnectChildren(adoptedChildren);
-}
-
-/**
- * The recursion doesn't need to be a complete traversal of the vnode graph,
- * instead it can be partial, when a custom element vnode is found, we don't
- * need to continue into its children because by attempting to disconnect the
- * custom element itself will trigger the removal of anything slotted or anything
- * defined on its shadow.
- */
-function recursivelyDisconnectChildren(vnodes: VNodes) {
-    for (let i = 0, len = vnodes.length; i < len; i += 1) {
-        const vnode: VCustomElement | VNode | null = vnodes[i];
-        if (!isNull(vnode) && isArray(vnode.children) && !isUndefined(vnode.elm)) {
-            // vnode is a VElement with children
-            if (isUndefined((vnode as any).ctor)) {
-                // it is a VElement, just keep looking (recursively)
-                recursivelyDisconnectChildren(vnode.children);
-            } else {
-                // it is a VCustomElement, disconnect it and ignore its children
-                resetComponentStateWhenRemoved(getCustomElementVM(vnode.elm as HTMLElement));
+    const { children, cmpRoot } = vm;
+    for (let i = 0, len = children.length; i < len; i += 1) {
+        const vnode = children[i];
+        if (!isNull(vnode)) {
+            const { elm } = vnode;
+            if (!isUndefined(elm)) {
+                // the concern here is that this routine might
+                // throw and we have one flow that does not
+                // have protection (recovering from error boundary)
```

Is this saying that the only time we expect this routine to throw is when we're recovering from an error boundary?

---

No, that is not what it says. 
It says that we are recovering from an error here and trying to clean up things. If removeChild() throws an error, its fine to ignore the error. For example, we were trying to clean up a vnode that was created but was never inserted to the shadow root, removeChild() will throw. And it is safe to ignore such an error.